### ISSUE TITLE
Wrong semantic of Dense layer for tf.python.keras.Dense when input has rank > 2.

### ISSUE THREAD

**Issue by fredo994** [tensorflow/tensorflow#25780]
**System information**
- OS Platform and Distribution: MacOs High Sierra
- TensorFlow installed from: installed from pip
- TensorFlow version: v1.12.0-0-ga6d8ffae09 1.12.0
- Python version: 3.6

**Problem**
I think that tf.python.keras.Dense layer is not implemented correctly. The documentation states:

>  Note: if the input to the layer has a rank greater than 2, then it is flattened prior to the initial dot product with `kernel`.

My interpretation of this claim is: if we have input tensor with shape `(None, 10, 10)` then dense layer with 20 hidden units should produce output with a shape `(None, 20)` and the kernel should have a shape of `(10 * 10, 20)`, and input tensor should be flattened to shape `(None, 100)` before product with kernel. 

**But**,  implementation is different. 

1. Kernel is initialized to shape `(10, 20)` (for the example above)
```python
self.kernel = self.add_weight(    # method Dense#build
        'kernel',
        shape=[input_shape[-1].value, self.units],
        initializer=self.kernel_initializer,
        regularizer=self.kernel_regularizer,
        constraint=self.kernel_constraint,
        dtype=self.dtype,
        trainable=True)
```

2. The output is computed as tensordot with summation over innermost dimensions
```python
# Broadcasting is required for the inputs.
outputs = standard_ops.tensordot(inputs, self.kernel, [[rank - 1], [0]]) # method Dense#call
```

This means that for `x.shape == (None, 10, 10)` and `units == 20` the output shape will be `(None, 10, 20)`, and that kernel is broadcasted for all other dimensions (this means weights are shared).

**Code to reproduce the issue**
```python
import tensorflow as tf
tf.enable_eager_execution()
x = tf.random.normal((10, 10, 10))
y = tf.layers.dense(x, 20)
print(y.shape)      # Produces (10, 10, 20) not (10, 20) as expected.
```

**Discussion**
Three possible things are happening here:

1. Documentation is wrong and implementation is correct, and a dense layer is meant to accept only rank 2-dimensional input (batchSize, a dimension of input vector) and if we pass it higher rank it treats it as a stack of input vectors. The solution is to change the documentation (specifically the note part), it should state something like numpy's documentation for `matmul`:  
> If either argument is N-D, N > 2, it is treated as a stack of matrices residing in the last two indexes and broadcast accordingly
2. Implementation is wrong and there should be `flatten` before `tensordot` (no need for `tensordot`, we can now use `matmul`)
3. I do not understand what does dense layer should do (unlikely because I asked my colleagues for their opinion and they agree with my view of what kind of operation does dense layer perform)

I've ordered this list from most to least likely (in my opinion).

**Comment by gshashank84** [comment]
I have added an comment for the working of **tensordot** function here, however this does not state that the Problem has been solved. 
#### The size of the specified input_shape still does not raise error if the shape of the inputs is not equal to input_shape.

**Comment by cofri** [comment]
@fredo994 I agree with your solution n°1 (remove or update the 'Note' part). 
I'm not an GitHub/PR expert, would you be able to open a PR with your proposition of changing the docstring?

**Comment by rmothukuru** [comment]
@fredo994,
A **Note**, as shown below, has been added in the [Documentation of Dense Layer](https://www.tensorflow.org/api_docs/python/tf/keras/layers/Dense):

> Note: If the input to the layer has a rank greater than 2, then Dense computes the dot product between the inputs and the kernel along the last axis of the inputs and axis 1 of the kernel (using tf.tensordot). For example, if input has dimensions (batch_size, d0, d1), then we create a kernel with shape (d1, units), and the kernel operates along axis 2 of the input, on every sub-tensor of shape (1, 1, d1) (there are batch_size * d0 such sub-tensors). The output in this case will have shape (batch_size, d0, units). 

Can you please confirm if we can close this issue, as it has been resolved? Thanks!
